keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md

```diff
@@ -0,0 +1,234 @@
+
+This architecture is similar to the approach taken by the exec based credential plugin architecture already present in client-go and CNI, and is a well understood pattern.  The API types are modeled after the ExecConfig and ExecCredential in client-go which define exec based credential retrieval for similar use cases.
+
+A RegistryCredentialProvider configuration API type (similar to [clientauthentication](https://github.com/kubernetes/kubernetes/tree/0273d43ae9486e9d0be292c01de2dd4143522b86/staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1)) will be added to Kubernetes:
+
+```go
+// RegistryCredentialProvider is used by the kubelet container runtime to match image URLs with
+// exec-based credential provider plugins that provide container registry credentials.
+type RegistryCredentialProvider struct {
+    metav1.TypeMeta `json:",inline"`
+
+    // ImageURLMatchers is a list of strings used to match against the image URL
+    // to determine which images to provide credentials for.  If one of the
+    // string matches the image URL, then the RegistryCredentialProvider will be
+    // used by kubelet to provide credentials for the image pull.
+    //
+    // Each ImageURLMatcher string is a URL pattern which can optionally contain
+    // a port and a path.  Globs can be used in the hostname (but not the port or
+    // the path).
+    //
+    // An image URL matches when it has the same number of parts as the
+    // ImageURLMatcher string, and each part matches.  Additionally the path of
+    // ImageURLMatcher must be a prefix of the target URL. If the ImageURLMatch
+    // contains a port, then the port must match as well.
+    ImageURLMatchers []string `json:"imageURLMatchers,omitempty"`
+
+    // Exec specifies a custom exec-based plugin.  This type is defined in
+    // https://github.com/kubernetes/client-go/blob/62f256057db7571c5ed1aba47eea291f72dd557a/tools/clientcmd/api/types.go#L184
+    Exec clientcmd.ExecConfig
+}
+```
+
+The RegistryCredentialProvider will be encoded in YAML and located in a file on disk.  The directory where credential provider configuration is found will be passed to kubelet via a new configuration option `RegistryCredentialProviderDir`.
+
+We will create new types `RegistryCredentialPluginRequest` and `RegistryCredentialPluginResponse` which will define the interface between the plugin and the kubelet runtime.  After the kubelet matches the image URL to a RegistryCredentialProvider, the kubelet will exec the plugin binary, and pass the JSON encoded request to the plugin via stdin.  This includes the image URL that is to be pulled.  The plugin will report back the response, which includes the credentials that kubelet needs to pull the image.
```

Comment:
If multiple providers are supported, which one "wins" if more than one matches?

Reply:
This is one area I wasn't quite sure about, but was going to try to mimic the current behavior as much as possible (I believe there is a "keyring" that is a union ImagePullSecrets with in-tree credential providers).  I can update the KEP when I get to this point during implementation, unless you think this is critical to decide up front.